File: src/main/java/frc/robot/subsystems/Climber.java
package frc.robot.subsystems;

import com.ctre.phoenix6.configs.CurrentLimitsConfigs;
import com.ctre.phoenix6.configs.TalonFXConfiguration;
import com.ctre.phoenix6.controls.VoltageOut;
import com.ctre.phoenix6.hardware.TalonFX;
import com.ctre.phoenix6.signals.InvertedValue;
import com.ctre.phoenix6.signals.NeutralModeValue;
import edu.wpi.first.wpilibj.DigitalInput;
import edu.wpi.first.wpilibj.motorcontrol.PWMTalonSRX;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import frc.robot.utils.RobotParameters;
import frc.robot.utils.RobotParameters.CoralManipulatorParameters;
import xyz.malefic.frc.pingu.AlertPingu;
import xyz.malefic.frc.pingu.VoltagePingu;

import static frc.robot.utils.RobotParameters.AlgaeManipulatorParameters.algaeIntaking;
import static frc.robot.utils.RobotParameters.ClimberParameters.CLIMBER_PINGU;
import static frc.robot.utils.RobotParameters.CoralManipulatorParameters.*;
import static frc.robot.utils.RobotParameters.MotorParameters.*;
import static frc.robot.utils.emu.CoralState.*;
import static xyz.malefic.frc.pingu.LogPingu.log;
import static xyz.malefic.frc.pingu.LogPingu.logs;

public class Climber extends SubsystemBase {
  private final TalonFX climbPivotMotor;
  private final PWMTalonSRX cageLockMotor;
  private final VoltageOut voltageOut;
  private final VoltageOut voltageOutFeeder;

  private final DigitalInput coralSensor;

  private boolean motorsRunning = false;

  /**
   * The Singleton instance of this CoralManipulatorSubsystem. Code should use the {@link
   * #getInstance()} method to get the single instance (rather than trying to construct an instance
   * of this class.)
   */
  private static final Climber INSTANCE = new Climber();

  /**
   * Returns the Singleton instance of this CoralManipulatorSubsystem. This static method should be
   * used, rather than the constructor, to get the single instance of this class. For example:
   * {@code CoralManipulatorSubsystem.getInstance();}
   */
  @SuppressWarnings("WeakerAccess")
  public static Climber getInstance() {
    return INSTANCE;
  }

  /**
   * Creates a new instance of this CoralManipulatorSubsystem. This constructor is private since
   * this class is a Singleton. Code should use the {@link #getInstance()} method to get the
   * singleton instance.
   */
  private Climber() {
    climbPivotMotor = new TalonFX(CLIMB_PIVOT_MOTOR_ID);
    cageLockMotor = new PWMTalonSRX(CAGE_LOCK_MOTOR_ID);

    coralSensor = new DigitalInput(CORAL_SENSOR_ID);

    TalonFXConfiguration climbPivotConfiguration = new TalonFXConfiguration();
//    PWNTalonSRXConfiguration cageLockConfiguration = new PWMTalonSRXConfiguration();

    climbPivotConfiguration.MotorOutput.NeutralMode = NeutralModeValue.Brake;

    climbPivotConfiguration.Slot0.kP = CLIMBER_PINGU.getP();
    climbPivotConfiguration.Slot0.kI = CLIMBER_PINGU.getI();
    climbPivotConfiguration.Slot0.kD = CLIMBER_PINGU.getD();
    climbPivotConfiguration.Slot0.kV = CLIMBER_PINGU.getV();

    climbPivotMotor.getConfigurator().apply(climbPivotConfiguration);

    CurrentLimitsConfigs climbPivotCurrentConfig = new CurrentLimitsConfigs();

    climbPivotCurrentConfig.SupplyCurrentLimit = 40;
    climbPivotCurrentConfig.SupplyCurrentLimitEnable = true;
    climbPivotCurrentConfig.StatorCurrentLimit = 40;
    climbPivotCurrentConfig.StatorCurrentLimitEnable = true;

    climbPivotMotor.getConfigurator().apply(climbPivotCurrentConfig);

    // on
    climbPivotConfiguration.MotorOutput.Inverted = InvertedValue.Clockwise_Positive;

    x.getConfigurator().apply(climbPivotConfiguration);

    voltageOut = new VoltageOut(0);
    voltageOutFeeder = new VoltageOut(0);

    AlertPingu.add(climbPivotMotor, "climber pivot motor");
  }

  /**
   * If the coral sensor is triggered, set the hasPiece boolean to true. (hasPiece = true,
   * sensorDetect = true), motors spinning If the manipulator has a piece, but the sensor no longer
   * detects it, stop the motors. (hasPiece = true, sensorDetect = false), motors stop If the
   * manipulator should start, but the motors are not running, start the motors (hasPiece = false,
   * sensorDetect = false), motors spinning by setting if it has a piece to false, due to the fact
   * that the manipulator should not have a piece after the motors are started again.
   *
   * <p>The manipulator motors should be on by default, as per Aaron's request.
   */
  @Override
  public void periodic() {




  }
`
  /** Stops the coral manipulator motors */
  public void stopClimbPivotMotor() {
    //    coralFeederMotor.stopMotor();
    climbPivotMotor.stopMotor();
  }

  /** Starts the coral manipulator motors */
  public void startCoralIntake() {
    voltageOut.Output = 5.0;
    coralScoreMotor.setControl(voltageOut);
    this.setHasPiece(false);
    //    isCoralIntaking = true;
  }

  /** Scores the coral motors */
  public void scoreCoral() {
    voltageOut.Output = 3.0;
    coralScoreMotor.setControl(voltageOut);
    this.motorsRunning = true;
    this.setHasPiece(false);
  }

  /** Starts the coral manipulator motors */
  public void slowCoralIntake() {
    coralScoreMotor.setControl(VoltagePingu.setOutput(4.0));
    //    coralFeederMotor.setControl(VoltagePingu.setOutput(4.0));
    this.setHasPiece(true);
  }

  /** Starts the coral manipulator motors */
  public void reverseMotors() {
    coralFeederMotor.setControl(VoltagePingu.setOutput(-4.5));
    coralScoreMotor.setControl(VoltagePingu.setOutput(-4.5));
    starFeederMotor.setControl(VoltagePingu.setOutput(-4.5));
    this.motorsRunning = true;
  }

  /**
   * Activates the algae intake process. This method stops the current motors, sets the voltage
   * output to 4.5, and starts the coral score motor to intake algae.
   */
  public void algaeIntake() {
    this.stopMotors();
    voltageOut.Output = -4.5;
    coralScoreMotor.setControl(voltageOut);
    this.motorsRunning = true;
  }

  /**
   * Sets the state of whether the manipulator has a piece.
   *
   * @param hasPiece true if the manipulator has a piece, false otherwise
   */
  public void setHasPiece(boolean hasPiece) {
    CoralManipulatorParameters.hasPiece = hasPiece;
  }

  /**
   * Sets the voltage output to -3.0 and controls the coral score motor to slow down the algae
   * scoring process.
   */
  public void slowAlgaeScoreMotors() {
    voltageOut.Output = -3.0;
    coralScoreMotor.setControl(voltageOut);
  }

  /**
   * Ejects the algae by setting the voltage output to 4.0 and controlling the coral score motor.
   */
  public void ejectAlgae() {
    voltageOut.Output = 4.0;
    coralScoreMotor.setControl(voltageOut);
  }

  /**
   * Gets the state of the coral manipulator
   *
   * @return The state of the coral manipulator
   */
  public boolean getCoralSensor() {
    return !coralSensor.get();
  }
}

File: src/main/java/frc/robot/subsystems/Climber.kt
package frc.robot.subsystems

import com.ctre.phoenix6.configs.ClosedLoopRampsConfigs
import com.ctre.phoenix6.configs.CurrentLimitsConfigs
import com.ctre.phoenix6.configs.MotionMagicConfigs
import com.ctre.phoenix6.configs.SoftwareLimitSwitchConfigs
import com.ctre.phoenix6.configs.TalonFXConfiguration
import com.ctre.phoenix6.controls.DutyCycleOut
import com.ctre.phoenix6.controls.MotionMagicVoltage
import com.ctre.phoenix6.controls.PositionDutyCycle
import com.ctre.phoenix6.controls.PositionVoltage
import com.ctre.phoenix6.controls.VelocityTorqueCurrentFOC
import com.ctre.phoenix6.controls.VoltageOut
import com.ctre.phoenix6.hardware.TalonFX
import com.ctre.phoenix6.signals.InvertedValue
import com.ctre.phoenix6.signals.NeutralModeValue
import edu.wpi.first.wpilibj.motorcontrol.PWMTalonSRX
import edu.wpi.first.wpilibj2.command.SubsystemBase
import frc.robot.utils.RobotParameters.ClimberParameters.PIVOT_MAGIC_PINGU
import frc.robot.utils.RobotParameters.ClimberParameters.PIVOT_PINGU
import frc.robot.utils.RobotParameters.ClimberParameters.PIVOT_SOFT_LIMIT_DOWN
import frc.robot.utils.RobotParameters.ClimberParameters.PIVOT_SOFT_LIMIT_UP
import frc.robot.utils.emu.ClimberPivotState
import org.littletonrobotics.junction.networktables.LoggedNetworkNumber
import xyz.malefic.frc.pingu.AlertPingu.add

object Climber : SubsystemBase() {
    private val pivotMotor = TalonFX(0)
    private val pivotConfigs: TalonFXConfiguration = TalonFXConfiguration()
    private val pivotSoftLimitConfig: SoftwareLimitSwitchConfigs
    private val posRequest: PositionDutyCycle
    private val velocityRequest: VelocityTorqueCurrentFOC
    private val voltageOut: VoltageOut
    private val motionMagicVoltage: MotionMagicVoltage
    private val cycleOut: DutyCycleOut
    private var motionMagicConfigs: MotionMagicConfigs
    private val voltagePos: PositionVoltage = PositionVoltage(0.0)
    private var pivotP: LoggedNetworkNumber? = null
    private var pivotI: LoggedNetworkNumber? = null
    private var pivotD: LoggedNetworkNumber? = null
    private var pivotV: LoggedNetworkNumber? = null
    private var pivotS: LoggedNetworkNumber? = null
    private var pivotG: LoggedNetworkNumber? = null
    private var cruiseV: LoggedNetworkNumber? = null
    private var acc: LoggedNetworkNumber? = null
    private var jerk: LoggedNetworkNumber? = null

    private var clampMotor: PWMTalonSRX = PWMTalonSRX(0)

    init {
        pivotConfigs.MotorOutput.NeutralMode = NeutralModeValue.Brake
        pivotConfigs.Slot0.kP = PIVOT_PINGU.p
        pivotConfigs.Slot0.kI = PIVOT_PINGU.i
        pivotConfigs.Slot0.kD = PIVOT_PINGU.d
        pivotConfigs.Slot0.kV = PIVOT_PINGU.v!!
        pivotConfigs.Slot0.kS = PIVOT_PINGU.s!!
        pivotMotor.configurator.apply(pivotConfigs)
        val pivotMotorCurrentConfig = CurrentLimitsConfigs()
        val pivotMotorRampConfig = ClosedLoopRampsConfigs()
        pivotSoftLimitConfig = SoftwareLimitSwitchConfigs()
        pivotMotorCurrentConfig.SupplyCurrentLimit = 40.79
        pivotMotorCurrentConfig.SupplyCurrentLimitEnable = true
        pivotMotor.configurator.apply(pivotMotorCurrentConfig)
        pivotMotorRampConfig.DutyCycleClosedLoopRampPeriod = 0.0
        pivotMotor.configurator.apply(pivotMotorRampConfig)
        pivotSoftLimitConfig.ForwardSoftLimitEnable = true
        pivotSoftLimitConfig.ReverseSoftLimitEnable = true
        pivotSoftLimitConfig.ForwardSoftLimitThreshold = PIVOT_SOFT_LIMIT_UP
        pivotSoftLimitConfig.ReverseSoftLimitThreshold = PIVOT_SOFT_LIMIT_DOWN
        pivotConfigs.MotorOutput.Inverted = InvertedValue.CounterClockwise_Positive
        pivotConfigs.SoftwareLimitSwitch = pivotSoftLimitConfig
        pivotConfigs.MotorOutput.DutyCycleNeutralDeadband = 0.1
        pivotMotor.configurator.apply(pivotSoftLimitConfig)
        velocityRequest = VelocityTorqueCurrentFOC(0.0)
        posRequest = PositionDutyCycle(0.0)
        voltageOut = VoltageOut(0.0)
        motionMagicVoltage = MotionMagicVoltage(0.0)
        cycleOut = DutyCycleOut(0.0)
        motionMagicConfigs = pivotConfigs.MotionMagic
        motionMagicConfigs.MotionMagicCruiseVelocity = PIVOT_MAGIC_PINGU.velocity
        motionMagicConfigs.MotionMagicAcceleration = PIVOT_MAGIC_PINGU.acceleration
        motionMagicConfigs.MotionMagicJerk = PIVOT_MAGIC_PINGU.jerk
        motionMagicVoltage.Slot = 0
        velocityRequest.OverrideCoastDurNeutral = false

        voltageOut.OverrideBrakeDurNeutral = false
        voltageOut.EnableFOC = true

        cycleOut.EnableFOC = false
        pivotMotor.setPosition(0.0)

        pivotMotor.configurator.apply(pivotConfigs)

        pivotMotor.configurator.apply(motionMagicConfigs)

        add(pivotMotor, "ClimberPivot")
        initializeLoggedNetworkPID()
    }
    fun initializeLoggedNetworkPID() {
        pivotP =
            LoggedNetworkNumber("Tuning/pivot/pivot P", pivotConfigs.Slot0.kP)
        pivotI =
            LoggedNetworkNumber("Tuning/pivot/pivot I", pivotConfigs.Slot0.kI)
        pivotD =
            LoggedNetworkNumber("Tuning/pivot/pivot D", pivotConfigs.Slot0.kD)
        pivotV =
            LoggedNetworkNumber("Tuning/pivot/pivot V", pivotConfigs.Slot0.kV)
        pivotS =
            LoggedNetworkNumber("Tuning/pivot/pivot S", pivotConfigs.Slot0.kS)
        pivotG =
            LoggedNetworkNumber("Tuning/pivot/pivot G", pivotConfigs.Slot0.kG)

        cruiseV =
            LoggedNetworkNumber(
                "Tuning/pivot/MM Cruise Velocity",
                motionMagicConfigs.MotionMagicCruiseVelocity,
            )
        acc =
            LoggedNetworkNumber(
                "Tuning/pivot/MM Acceleration",
                motionMagicConfigs.MotionMagicAcceleration,
            )
        jerk = LoggedNetworkNumber("Tuning/pivot/MM Jerk", motionMagicConfigs.MotionMagicJerk)

    }
    fun setPivotPos(state: ClimberPivotState) {
        pivotMotor.setControl(voltagePos.withPosition(state.pos))
    }
    fun closeClampMotor() {
        clampMotor.set(-1.0);
    }
    fun openClampMotor() {
        clampMotor.set(1.0);
    }
    fun speedClamp(speed: Double) {
        clampMotor.set(speed)
    }
}